src/js/ripple/keypairs.js

```diff
@@ -0,0 +1,283 @@
+@param {Object|String} specifier - the specifier for the KeyPair
+ */
+function KeyPair(seed, specifier) {
+  this.seed = seed;
+  this.specifier = specifier;
+}
+
+/*
+@param {Array} message
+@virtual
+ */
+KeyPair.prototype.sign = isVirtual;
+
+
+/*
+@param {Array<Number>} message
+@param {Array<Number>} signature
+@virtual
+ */
+KeyPair.prototype.verify = isVirtual;
+
+/*
+@return {Array<Number>} of bytes, in canonical form, for signing
+@virtual
+*/
+KeyPair.prototype.pubKeyBytes = isVirtual;
+
+hasCachedProperty(KeyPair.prototype, 'pubKeyHex', function() {
+  return bytesToHex(this.pubKeyBytes());
+});
+
+hasCachedProperty(KeyPair.prototype, 'account', function() {
+  return createAccountId(this.pubKeyBytes());
+});
+
+KeyPair.prototype.signHex = function(message) {
+  return bytesToHex(this.sign(message));
+};
+
+/* ---------------------------- SECP256K1 KEYPAIR --------------------------- */
+
+/*
+* @class
+* @private
+*/
+function Ed25519Pair() {
+  KeyPair.apply(this, arguments);
+  var secretBytes = deriveEdKeyPairSeed(this.seed);
+  var keyPair = nacl.sign.keyPair.fromSeed(secretBytes);
+  this.secretKey = keyPair.secretKey;
+  this.publicKey = keyPair.publicKey;
+  this.type = KeyType.ed25519;
+}
+
+util.inherits(Ed25519Pair, KeyPair);
+
+hasCachedProperty(Ed25519Pair.prototype, 'pubKeyBytes', function() {
+  return [0xED].concat(toGenericArray(this.publicKey));
+});
+
+Ed25519Pair.prototype.sign = function(message) {
+  var messageArray = new Uint8Array(message);
+  var signatureArray = nacl.sign.detached(messageArray, this.secretKey);
+  return toGenericArray(signatureArray);
+};
+
+Ed25519Pair.prototype.verify = function(message, signature) {
+  var messageArray = new Uint8Array(message);
+  var signatureArray = new Uint8Array(signature);
+  return nacl.sign.detached.verify(messageArray, signatureArray,
+                                   this.publicKey);
+};
+
+/* ---------------------------- SECP256K1 KEYPAIR --------------------------- */
+
+/*
+* @class
+* @private
+*/
+function Secp256k1Pair() {
+  KeyPair.apply(this, arguments);
+  this.keyPair = this.seed.get_key();
+  this.type = KeyType.secp256k1;
+}
+
+util.inherits(Secp256k1Pair, KeyPair);
+
+hasCachedProperty(Secp256k1Pair.prototype, 'pubKeyBytes', function() {
+  return hexToArray(this.keyPair.to_hex_pub());
+});
+
+/*
+@param {Array<Number>} message (bytes)
+ */
+Secp256k1Pair.prototype.sign = function(message) {
+  var hash = this.hashMessage(message);
+  var signatureBits = this.keyPair.sign(hash, 0);
+  return sjcl.codec.bytes.fromBits(signatureBits);
+};
+
+/*
+@param {Array<Number>} message (bytes)
+@return {UInt256} 256 bit hash of the message
+ */
+Secp256k1Pair.prototype.hashMessage = function(message) {
+  var bits = sjcl.codec.bytes.toBits(message);
+  return Crypt.hashSha512Half(bits);
+};
+
+/*
+@param {Array<Number>} message (bytes)
+@param {Array<Number>} signature (bytes)
+ */
+Secp256k1Pair.prototype.verify = function(message, signature) {
```

Comment:
This logic seems to be extracting r and s from a DER encoded signature. I don't think ed25519 uses DER encoding for sigs.